### PR TITLE
[extractor/tv4] Fix tv4 extraction

### LINKED ISSUE #5535: [TV4] HTTP Error 503: Service Unavailable
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting a broken site
- [X] I've verified that I'm running yt-dlp version **2022.11.11** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/yt-dlp/yt-dlp/wiki/FAQ#video-url-contains-an-ampersand--and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

Sweden

### Provide a description that is worded well enough to be understood

www.tv4play.se not working:
Unable to download JSON metadata: HTTP Error 503: Service Unavailable

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
[debug] Command-line config: ['https://www.tv4play.se/program/next-level-chef-us/del-5-s%C3%A4song-1/13772514', '-vU']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.11.11 [8b64402] (pip)
[debug] Python 3.10.6 (CPython x86_64 64bit) - Linux-5.15.0-52-generic-x86_64-with-glibc2.35 (OpenSSL 3.0.2 15 Mar 2022, glibc 2.35)
[debug] exe versions: ffmpeg 4.4.2 (setts), ffprobe 4.4.2, rtmpdump 2.4
[debug] Optional libraries: Cryptodome-3.11.0, brotli-1.0.9, certifi-2020.06.20, mutagen-1.46.0, pyxattr-0.7.2, secretstorage-3.3.1, sqlite3-2.6.0, websockets-10.4
[debug] Proxy map: {}
[debug] Loaded 1723 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
Latest version: 2022.11.11, Current version: 2022.11.11
yt-dlp is up to date (2022.11.11)
[debug] Using fake IP 78.67.182.65 (SE) as X-Forwarded-For
[debug] [TV4] Extracting URL: https://www.tv4play.se/program/next-level-chef-us/del-5-s%C3%A4song-1/13772514
[TV4] 13772514: Downloading video info JSON
ERROR: [TV4] 13772514: Unable to download JSON metadata: HTTP Error 503: Service Unavailable (caused by <HTTPError 503: 'Service Unavailable'>); please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U
  File "/home/user/.local/lib/python3.10/site-packages/yt_dlp/extractor/common.py", line 674, in extract
    ie_result = self._real_extract(url)
  File "/home/user/.local/lib/python3.10/site-packages/yt_dlp/extractor/tv4.py", line 75, in _real_extract
    info = self._download_json(
  File "/home/user/.local/lib/python3.10/site-packages/yt_dlp/extractor/common.py", line 1034, in download_content
    res = getattr(self, download_handle.__name__)(url_or_request, video_id, **kwargs)
  File "/home/user/.local/lib/python3.10/site-packages/yt_dlp/extractor/common.py", line 998, in download_handle
    res = self._download_webpage_handle(
  File "/home/user/.local/lib/python3.10/site-packages/yt_dlp/extractor/common.py", line 868, in _download_webpage_handle
    urlh = self._request_webpage(url_or_request, video_id, note, errnote, fatal, data=data, headers=headers, query=query, expected_status=expected_status)
  File "/home/user/.local/lib/python3.10/site-packages/yt_dlp/extractor/common.py", line 825, in _request_webpage
    raise ExtractorError(errmsg, cause=err)

  File "/home/user/.local/lib/python3.10/site-packages/yt_dlp/extractor/common.py", line 807, in _request_webpage
    return self._downloader.urlopen(self._create_request(url_or_request, data, headers, query))
  File "/home/user/.local/lib/python3.10/site-packages/yt_dlp/YoutubeDL.py", line 3692, in urlopen
    return self._opener.open(req, timeout=self._socket_timeout)
  File "/usr/lib/python3.10/urllib/request.py", line 525, in open
    response = meth(req, response)
  File "/usr/lib/python3.10/urllib/request.py", line 634, in http_response
    response = self.parent.error(
  File "/usr/lib/python3.10/urllib/request.py", line 563, in error
    return self._call_chain(*args)
  File "/usr/lib/python3.10/urllib/request.py", line 496, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.10/urllib/request.py", line 643, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 503: Service Unavailable
```

### PR DESCRIPTION
**Pull request by TxI5**
changed json api url

**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

</details>

TV4 has changed their api domain. This changes the url for json metadata extraction

Fixes #5535 

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [ ] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

**Comment by TxI5** [comment]
Closes #5535

**Comment by dirkf** [comment]
Although not related to the original issue, the geo-restriction processing could be also improved. XFF doesn't work for this site.

I've applied some yt-dl test code that could be adapted:
```diff
--- old/yt_dlp/extractor/tv4.py
+++ new/yt_dlp/extractor/tv4.py
@@ -20,7 +20,7 @@
                 sport/|
             )
         )(?P<id>[0-9]+)'''
-    _GEO_COUNTRIES = ['SE']
+    _GEO_COUNTRIES = False
     _TESTS = [
         {
             'url': 'http://www.tv4.se/kalla-fakta/klipp/kalla-fakta-5-english-subtitles-2491650',
@@ -73,13 +73,23 @@
         video_id = self._match_id(url)
 
         info = self._download_json(
-            'https://playback-api.b17g.net/asset/%s' % video_id,
+            'https://playback2.a2d.tv/play/' + video_id,
             video_id, 'Downloading video info JSON', query={
                 'service': 'tv4',
                 'device': 'browser',
                 'protocol': 'hls,dash',
                 'drm': 'widevine',
-            })['metadata']
+            }, expected_status=401)
+        err = info.get('errorCode')
+        if err:
+            msg = info.get('message') or err
+            if 'GEO_LOCATION' in err:
+                self.report_warning(msg, video_id)
+                self.raise_geo_restricted(countries=['SE'])
+            err = compat_HTTPError('https://playback2.a2d.tv/play/' + video_id, 401, 'HTTP Error 401: Unauthorized', None, None)
+            raise ExtractorError(msg, cause=error_to_compat_str(err), video_id=video_id)
+
+        info = info['metadata']
 
         title = info['title']
 
```

File: yt_dlp/extractor/tv4.py [hunk 1]
```diff
@@ -73,7 +73,7 @@ def _real_extract(self, url):
         video_id = self._match_id(url)
 
         info = self._download_json(
-            'https://playback-api.b17g.net/asset/%s' % video_id,
+            'https://playback2.a2d.tv/asset/%s' % video_id,
```
Comment (bashonly) [inline yt_dlp/extractor/tv4.py:76]: use f-strings for both API URLs for consistency / readability
```suggestion
            f'https://playback2.a2d.tv/asset/{video_id}',
```

File: yt_dlp/extractor/tv4.py [hunk 2]
```diff
@@ -84,7 +84,7 @@ def _real_extract(self, url):
         title = info['title']
 
         manifest_url = self._download_json(
-            'https://playback-api.b17g.net/media/' + video_id,
+            'https://playback2.a2d.tv/play/' + video_id,
```
Comment (bashonly) [inline yt_dlp/extractor/tv4.py:87]:
```suggestion
            f'https://playback2.a2d.tv/play/{video_id}',
```

**Comment by bashonly** [comment]
> Although not related to the original issue, the geo-restriction processing could be also improved. XFF doesn't work for this site.
> 
> I've applied some yt-dl test code that could be adapted:

This is a good idea, since it seems like all tv4 content is geo-restricted now. But the diff posted above conflates the `asset` endpoint with the `play` endpoint.

A working version could look like this (using this PR's initial commit as baseline for the diff):
```diff
-        manifest_url = self._download_json(
-            'https://playback2.a2d.tv/play/' + video_id,
-            video_id, query={
+        manifest_info = self._download_json(
+            f'https://playback2.a2d.tv/play/{video_id}',
+            video_id, 'Downloading manifest info JSON', query={
                 'service': 'tv4',
                 'device': 'browser',
                 'protocol': 'hls',
-            })['playbackItem']['manifestUrl']
+            }, expected_status=401)
+        err = manifest_info.get('errorCode')
+        if err:
+            msg = manifest_info.get('message') or err
+            if 'GEO_LOCATION' in err:
+                self.raise_geo_restricted(msg, countries=['SE'])
+            raise ExtractorError(f'HTTP Error 401: Unauthorized; {msg}', video_id=video_id)
+
+        manifest_url = manifest_info['playbackItem']['manifestUrl']
+
```

File: yt_dlp/extractor/tv4.py (hunk 1 above)
Comment (dirkf) [reply yt_dlp/extractor/tv4.py:76]: Definitely make them consistent, but if it's done using f-strings yt-dl will just have to change it to any of the other myriad string syntaxes (I know that's our burden).

Maybe s/t like this:
```py
    def _call_api(self, endpoint, video_id, query=None, **kwargs):
        return self._download_json(
            '/'.join(('https://playback2.a2d.tv', endpoint, video_id)),
            video_id, endpoint.join(('Downloading video ', ' JSON')),
            query=update_url_query({
                'service': 'tv4',
                'device': 'browser',
                'protocol': 'hls,dash',
            }, query), **kwargs)
...
        info = self._call_api('asset', video_id, query={
                'drm': 'widevine',
            })['metadata']
...
        manifest_url = self._call_api('play', video_id, expected_status=401)
        # don't crash during error handling
        err = traverse_obj(manifest_url, 'errorCode')
...
        # then as in https://github.com/yt-dlp/yt-dlp/pull/5649#issuecomment-1337847672 below
```
Related to the geo-restriction handling, in the class:
```py
    # XFF is not effective
    _GEO_BYPASS = False
```
and perhaps specialise the `msg` for `raise_geo_restricted()` if, as I believe, a logged-in user can access the site from outside SE/EU:
```py
                self.raise_geo_restricted(
                    'This video is not available from your location due to geo restriction, or not being authenticated.',
                    countries=['SE'])
```

Comment (pukkandan) [reply yt_dlp/extractor/tv4.py:76]: > Definitely make them consistent, but if it's done using f-strings yt-dl will just have to change it to any of the other myriad string syntaxes (I know that's our burden).

Sorry, but if we discourage the use of newer Python features, then there was no point in deprecating older versions. In most cases (including here) fstrings simply read much better than the alternatives.

You could encourage OP to make PR against youtube-dl and get it merged there first - which would reverse the burdens; i.e. the burden of code review will then mainly fall on you, and the burden of porting (dealing with deprecated imports/functions) on me

Comment (dirkf) [reply yt_dlp/extractor/tv4.py:76]: Fair point, but in this case it is at least possible to have just the one instance, and also avoid replicating the base API URL.

Comment (pukkandan) [reply yt_dlp/extractor/tv4.py:76]: oh, I wasn't commenting on the whole thing, just on the quoted part. Having a `_call_api` does look good

**Comment by pukkandan** [comment]
Does this still work? Pls update tests

**Review by bashonly — APPROVED** [review]

File: yt_dlp/extractor/tv4.py [hunk 3]
```diff
@@ -20,19 +23,25 @@ class TV4IE(InfoExtractor):
                 sport/|
             )
         )(?P<id>[0-9]+)'''
-    _GEO_COUNTRIES = ['SE']
+    _GEO_BYPASS = False
```
Comment (pukkandan) [inline yt_dlp/extractor/tv4.py:26]: Why?

**Review by pukkandan — APPROVED** [review]

File: yt_dlp/extractor/tv4.py (hunk 3 above)
Comment (bashonly) [reply yt_dlp/extractor/tv4.py:26]: per dirkf's comment here https://github.com/yt-dlp/yt-dlp/pull/5649#discussion_r1039935656 that xff is futile

Comment (pukkandan) [reply yt_dlp/extractor/tv4.py:26]: Then why r we passing go verification headers? Am I misunderstanding something?

Comment (bashonly) [reply yt_dlp/extractor/tv4.py:26]: `_GEO_BYPASS` is for `-xff`, which has no effect
`geo_verification_headers()` are for `--geo-verification-proxy`, which does work

**Pull request merged**